hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotCache.java

```diff
@@ -0,0 +1,300 @@
+    FS_API_READ,
+    SNAPDIFF_READ,
+    DEEP_CLEAN_WRITE,
+    GARBAGE_COLLECTION_WRITE
+  }
+
+  /**
+   * Get or load OmSnapshot. Must be close()d after use.
+   * TODO: [SNAPSHOT] Can add reason enum to param list later.
+   * @param key snapshot table key
+   * @return an OmSnapshot instance, or null on error
+   */
+  public ReferenceCounted<IOmMetadataReader> get(String key)
+      throws IOException {
+    // Atomic operation to initialize the OmSnapshot instance (once) if the key
+    // does not exist.
+    ReferenceCounted<IOmMetadataReader> rcOmSnapshot =
+        dbMap.computeIfAbsent(key, k -> {
+          LOG.info("Loading snapshot. Table key: {}", k);
+          try {
+            return new ReferenceCounted<>(cacheLoader.load(k));
+          } catch (OMException omEx) {
+            // Return null if the snapshot is no longer active
+            if (!omEx.getResult().equals(FILE_NOT_FOUND)) {
+              throw new IllegalStateException(omEx);
+            }
+          } catch (IOException ioEx) {
+            // Failed to load snapshot DB
+            throw new IllegalStateException(ioEx);
+          } catch (Exception ex) {
+            // Unexpected and unknown exception thrown from CacheLoader#load
+            throw new IllegalStateException(ex);
+          }
+          // Do not put the value in the map on exception
+          return null;
+        });
+
+    if (rcOmSnapshot == null) {
+      // The only exception that would fall through the loader logic above
+      // is OMException with FILE_NOT_FOUND.
+      throw new OMException("Snapshot table key '" + key + "' not found, "
+          + "or the snapshot is no longer active",
+          OMException.ResultCodes.FILE_NOT_FOUND);
+    }
+
+    // If the snapshot is already loaded in cache, the check inside the loader
+    // above is ignored. But we would still want to reject all get()s except
+    // when called from SDT (and some) if the snapshot is not active any more.
+    if (!omSnapshotManager.isSnapshotStatus(key, SNAPSHOT_ACTIVE) &&
+        !SnapshotUtils.isCalledFromSnapshotDeletingService()) {
+      throw new OMException("Unable to load snapshot. " +
+          "Snapshot with table key '" + key + "' is no longer active",
+          FILE_NOT_FOUND);
+    }
+
+    // Increment the reference count on the instance.
+    rcOmSnapshot.incrementRefCount();
+
+    // Remove instance from clean up list when it exists.
+    // TODO: [SNAPSHOT] Check thread safety with release()
+    pendingEvictionList.remove(rcOmSnapshot);
+
+    // Check if any entries can be cleaned up.
+    // At this point, cache size might temporarily exceed cacheSizeLimit
+    // even if there are entries that can be evicted, which is fine since it
+    // is a soft limit.
+    cleanup();
+
+    return rcOmSnapshot;
+  }
+
+  /**
+   * Release the reference count on the OmSnapshot instance.
+   * @param key snapshot table key
+   */
+  public void release(String key) {
```

Comment:
I only see this method being called in the tests.  Doesn't it need to be called from the real code somewhere?

Reply:
Hi @GeorgeJahad , this method is not directly called from prod code anymore because we use `try-with-resources` now. It is only called internally. I think I can keep it here since `LoadingCache` has it. (Or mark it private)

Reply:
Thanks @smengcl   I see that try-with-resources is calling ReferenceCounted::close().  My question is how does that invoke SnapshotCache::release()?

Reply:
Ah I see, `ReferenceCounted::close` does not directly trigger `SnapshotCache::release` right now.

I would need to incorporate the `pendingEvictionList` logic in there somehow. So that `cleanup()` would act correctly upon cache eviction.

Reply:
Maybe add a close() to SnapshotCache and make ReferenceCounted<> only take closeables?:
```
public class ReferenceCounted<T implements Closeable> implements AutoCloseable {
```

Reply:
ah close() doesn't take a parmeter, so it would need to be a different interface, but same principle.

Reply:
`ReferenceCounted<>` are inside a map in `SnapshotCache`. So that would require me to pass the reference to `SnapshotCache` every time I create a new `ReferenceCounted<>` right?

Is there a better way?

Reply:
You are right, I wasn't thinking about it correctly.

Reply:
No worries. :D

The native approach would be to O(1) scan the `dbMap` every time `cleanup()` is called, but that obviously wouldn't be ideal.

Only if there is a `ConcurrentTreeMap` that can sort based on **values**.

Or I could bake a doubly linked list inside `ReferenceCounted` so each `ReferenceCounted<>` instance becomes a node..

Reply:
To me, the object passed into the `ReferenceCount<>` constructor should have a notion of how to clean up after itself after the refcount goes to 0.

That seems like a fundamental part of a generic reference counting algorithm.  It shouldn't know about the object itself, just that the object has a particular interface.

I guess I'm basically going back to my original thought.  One possiblity is to add a default no-op close method to `IOmMetadataReader` interface.  Then have the `SnapshotCache.get()` subclass that to return an `IOmMetadataReader` object which delegates most calls to the internal `IOmMetadataReader` but also has a close method that calls this.release().

And then make `ReferenceCounted<>` only take closeables like I mentioned earlier:
```
public class ReferenceCounted<T implements Closeable> implements AutoCloseable {
```

and invoke obj.close() from within the `ReferenceCounted::close`

Reply:
Actually, the `OzoneManager`  isalso an `IOmMetadataReader` that implements close(), which we don't want to accidentally invoke here.  So instead of using the `Closable` interface here, I would probably create a new different interface, "Releaseable" and have `ReferenceCounted` take that:
```
public class ReferenceCounted<T implements Releaseable> implements AutoCloseable {
```

Reply:
@GeorgeJahad 

That shouldn't be a problem with the current implementation, we are calling `close()` on a different level. Even if `ReferenceCounted#close` is called when the wrapped `IOmMetadataReader` is an `OzoneManager` instance, it is only decrementing the reference count (if RC is enabled on active DB).

Because active DB will never be in `SnapshotCache`, it is nevered queued in `pendingEvictionList`. Thus, `OzoneManager#close` will never be called in the case of RC.